File: src/Runtime/mmdRuntimeBone.ts
import { Matrix, Quaternion, Vector3 } from "@babylonjs/core/Maths/math.vector";
import type { DeepImmutable, Nullable } from "@babylonjs/core/types";

import type { MmdModelMetadata } from "@/Loader/mmdModelMetadata";
import { PmxObject } from "@/Loader/Parser/pmxObject";

import type { AppendTransformSolver } from "./appendTransformSolver";
import { quaternionToAxisAngle } from "./Common/quaternionToAxisAngle";
import type { IkChainInfo } from "./ikChainInfo";
import type { IkSolver } from "./ikSolver";
import type { IMmdRuntimeBone } from "./IMmdRuntimeBone";
import type { IMmdRuntimeLinkedBone } from "./IMmdRuntimeLinkedBone";

/**
 * Bone for MMD runtime
 *
 * For mmd runtime, it is necessary to override the bone system because it has a different implementation than the usual matrix update method
 *
 * Which requires the mmd runtime bone, which is the wrapper of the babylon.js bone
 */
export class MmdRuntimeBone implements IMmdRuntimeBone {
    /**
     * The Babylon.js bone
     */
    public readonly linkedBone: IMmdRuntimeLinkedBone;

    /**
     * Name of the bone
     */
    public readonly name: string;

    /**
     * Parent bone
     */
    public parentBone: Nullable<MmdRuntimeBone>;

    /**
     * Child bones
     */
    public readonly childBones: MmdRuntimeBone[];

    /**
     * Transform order
     */
    public readonly transformOrder: number;

    /**
     * Bone flag
     *
     * @see PmxObject.Bone.Flag
     */
    public readonly flag: number;

    /**
     * Whether the bone transform is applied after physics
     */
    public readonly transformAfterPhysics: boolean;

    /**
     * Append transform solver
     *
     * If the bone does not have an append transform solver, it will be null
     */
    public appendTransformSolver: Nullable<AppendTransformSolver>;

    /**
     * Axis limitation of the bone
     */
    public readonly axisLimit: Nullable<DeepImmutable<Vector3>>;

    /**
     * IK solver
     *
     * If the bone does not have an ik solver, it will be null
     */
    public ikSolver: Nullable<IkSolver>;

    /**
     * The position offset value to be moved by the bone morph
     *
     * This is a field that is primarily updated by the morph controller
     */
    public readonly morphPositionOffset: Vector3;

    /**
     * The rotation offset value to be moved by the bone morph
     *
     * This is a field that is primarily updated by the morph controller
     */
    public readonly morphRotationOffset: Quaternion;

    /**
     * IK chain bone states
     *
     * If this bone is an Ik chain, this value is non-null
     */
    public ikChainInfo: Nullable<IkChainInfo>;

    /**
     * World matrix of this bone
     *
     * Slice of `MmdModel.worldTransformMatrices` that corresponds to this bone
     */
    public readonly worldMatrix: Float32Array;

    // /**
    //  * Gets the position of a local transform with animations and bone morphs applied
    //  */
    // public getAnimatedPositionToRef: (target: Vector3) => Vector3;

    /**
     * Gets the rotation of a local transform with animations and bone morphs applied
     */
    public getAnimatedRotationToRef: (target: Quaternion) => Quaternion;

    /**
     * Get the position offset of the local transform with animation and bone morph applied
     *
     * Refers to the change from the rest position
     */
    public getAnimationPositionOffsetToRef: (target: Vector3) => Vector3;
    // public getAnimationRotationOffsetToRef: (target: Quaternion) => Quaternion;

    /**
     * Create MMD runtime bone
     * @param linkedBone Linked Babylon.js bone
     * @param boneMetadata Bone metadata
     * @param worldTransformMatrices World transform matrices
     * @param boneIndex Bone index
     */
    public constructor(
        linkedBone: IMmdRuntimeLinkedBone,
        boneMetadata: MmdModelMetadata.Bone,
        worldTransformMatrices: Float32Array,
        boneIndex: number
    ) {
        this.linkedBone = linkedBone;

        this.name = boneMetadata.name;
        this.parentBone = null;
        this.childBones = [];

        this.transformOrder = boneMetadata.transformOrder;
        this.flag = boneMetadata.flag;
        this.transformAfterPhysics = (boneMetadata.flag & PmxObject.Bone.Flag.TransformAfterPhysics) !== 0;

        this.appendTransformSolver = null;
        if (boneMetadata.axisLimit !== undefined) {
            this.axisLimit = Vector3.FromArray(boneMetadata.axisLimit).normalize();
        } else {
            this.axisLimit = null;
        }
        this.ikSolver = null;

        this.morphPositionOffset = Vector3.Zero();
        this.morphRotationOffset = Quaternion.Identity();

        this.ikChainInfo = null;

        this.worldMatrix = new Float32Array(worldTransformMatrices.buffer, worldTransformMatrices.byteOffset + boneIndex * 16 * 4, 16);

        // this.getAnimatedPositionToRef = this._getAnimatedPositionToRef;
        this.getAnimatedRotationToRef = this._getAnimatedRotationToRef;
        this.getAnimationPositionOffsetToRef = this._getAnimationPositionOffsetToRef;
        // this.getAnimationRotationOffsetToRef = this._getAnimationRotationOffsetToRef;
    }

    // private _getAnimatedPositionWithMorphToRef(target: Vector3): Vector3 {
    //     target.copyFrom(this.linkedBone.position);
    //     return target.addInPlace(this.morphPositionOffset);
    // }

    // private _getAnimatedPositionToRef(target: Vector3): Vector3 {
    //     target.copyFrom(this.linkedBone.position);
    //     return target;
    // }

    private static readonly _TempAxis = new Vector3();

    private _getAnimatedRotationToRef(target: Quaternion): Quaternion {
        // MMD's implementation transforms the rotation axis to fit the axis limit of the target skeleton at animation load time.
        // However, that method makes it impossible to apply one animation data to multiple models,
        // so we use an implementation that performs the axis transformation at runtime.
        if (this.axisLimit !== null) {
            const animationAxis = MmdRuntimeBone._TempAxis;
            let angle = quaternionToAxisAngle(this.linkedBone.rotationQuaternion, animationAxis);
            if (Vector3.Dot(animationAxis, this.axisLimit) < 0) {
                angle = -angle;
            }
            Quaternion.RotationAxisToRef(this.axisLimit, angle, target);
        } else {
            target.copyFrom(this.linkedBone.rotationQuaternion);
        }
        return target;
    }

    private _getAnimatedRotationWithMorphToRef(target: Quaternion): Quaternion {
        if (this.axisLimit !== null) {
            const animationAxis = MmdRuntimeBone._TempAxis;
            let angle = quaternionToAxisAngle(this.linkedBone.rotationQuaternion, animationAxis);
            if (Vector3.Dot(animationAxis, this.axisLimit) < 0) {
                angle = -angle;
            }
            Quaternion.RotationAxisToRef(this.axisLimit, angle, target);
        } else {
            target.copyFrom(this.linkedBone.rotationQuaternion);
        }
        return this.morphRotationOffset.multiplyToRef(target, target);
    }

    private static readonly _TempVector3 = new Vector3();

    private _getAnimationPositionOffsetToRef(target: Vector3): Vector3 {
        target.copyFrom(this.linkedBone.position);
        this.linkedBone.getRestMatrix().getTranslationToRef(MmdRuntimeBone._TempVector3);
        return target.subtractInPlace(MmdRuntimeBone._TempVector3);
    }

    private _getAnimationPositionOffsetWithMorphToRef(target: Vector3): Vector3 {
        target.copyFrom(this.linkedBone.position);
        target.addInPlace(this.morphPositionOffset);
        this.linkedBone.getRestMatrix().getTranslationToRef(MmdRuntimeBone._TempVector3);
        return target.subtractInPlace(MmdRuntimeBone._TempVector3);
    }

    // a: rest quaternion
    // b: animation quaternion
    // c: animated quaternion

    // a * b = c

    // to get b from a and c:
    // a^-1 * c = b

    // private static readonly _TempQuaternion = new Quaternion();

    // private _getAnimationRotationOffsetToRef(target: Quaternion): Quaternion {
    //     target.copyFrom(this.linkedBone.rotationQuaternion);
    //     Quaternion.FromRotationMatrixToRef(this.linkedBone.getRestMatrix(), MmdRuntimeBone._TempQuaternion).invertInPlace();
    //     return MmdRuntimeBone._TempQuaternion.multiplyInPlace(target);
    // }

    // private _getAnimationRotationOffsetWithMorphToRef(target: Quaternion): Quaternion {
    //     target.copyFrom(this.linkedBone.rotationQuaternion);
    //     this.morphRotationOffset.multiplyToRef(target, target);
    //     Quaternion.FromRotationMatrixToRef(this.linkedBone.getRestMatrix(), MmdRuntimeBone._TempQuaternion).invertInPlace();
    //     return MmdRuntimeBone._TempQuaternion.multiplyInPlace(target);
    // }

    /**
     * Allows the animation of this bone to be affected by the `morphPositionOffset` and `morphRotationOffset` fields
     */
    public enableMorph(): void {
        // this.getAnimatedPositionToRef = this._getAnimatedPositionWithMorphToRef;
        this.getAnimatedRotationToRef = this._getAnimatedRotationWithMorphToRef;

        this.getAnimationPositionOffsetToRef = this._getAnimationPositionOffsetWithMorphToRef;
        // this.getAnimationRotationOffsetToRef = this._getAnimationRotationOffsetWithMorphToRef;
    }

    /**
     * Disables the animation of this bone to be affected by the `morphPositionOffset` and `morphRotationOffset` fields
     */
    public disableMorph(): void {
        // this.getAnimatedPositionToRef = this._getAnimatedPositionToRef;
        this.getAnimatedRotationToRef = this._getAnimatedRotationToRef;

        this.getAnimationPositionOffsetToRef = this._getAnimationPositionOffsetToRef;
        // this.getAnimationRotationOffsetToRef = this._getAnimationRotationOffsetToRef;
    }

    /**
     * Reset world matrix, append transform, and ik chain state
     * @internal
     */
    public resetTransformState(): void {
        const worldMatrix = this.worldMatrix;
        worldMatrix[0] = 1;
        worldMatrix[1] = 0;
        worldMatrix[2] = 0;
        worldMatrix[3] = 0;

        worldMatrix[4] = 0;
        worldMatrix[5] = 1;
        worldMatrix[6] = 0;
        worldMatrix[7] = 0;

        worldMatrix[8] = 0;
        worldMatrix[9] = 0;
        worldMatrix[10] = 1;
        worldMatrix[11] = 0;

        worldMatrix[12] = 0;
        worldMatrix[13] = 0;
        worldMatrix[14] = 0;
        worldMatrix[15] = 1;

        if (this.ikChainInfo !== null) {
            this.ikChainInfo.localRotation.set(0, 0, 0, 1);
            this.ikChainInfo.localPosition.set(0, 0, 0);
            this.ikChainInfo.ikRotation.set(0, 0, 0, 1);
        }

        if (this.appendTransformSolver !== null) {
            this.appendTransformSolver.appendRotation.set(0, 0, 0, 1);
            this.appendTransformSolver.appendPosition.set(0, 0, 0);
        }
    }

    private static readonly _TempRotation = Quaternion.Identity();
    private static readonly _TempPosition = Vector3.Zero();
    private static readonly _TempPosition2 = Vector3.Zero();
    private static readonly _TempMatrix = Matrix.Identity();
    private static readonly _TempMatrix2 = Matrix.Identity();

    /**
     * Update the world matrix of this bone to account for append transform and ik
     * @param usePhysics Whether to use physics simulation
     * @param computeIk Whether to compute ik
     * @internal
     */
    public updateWorldMatrix(usePhysics: boolean, computeIk: boolean): void {
        const rotation = this.getAnimatedRotationToRef(MmdRuntimeBone._TempRotation);
        const position = this.getAnimationPositionOffsetToRef(MmdRuntimeBone._TempPosition);

        if (this.appendTransformSolver !== null) {
            this.appendTransformSolver.update(rotation, position);

            if (this.appendTransformSolver.affectRotation) {
                rotation.copyFrom(this.appendTransformSolver.appendRotation);
            }

            if (this.appendTransformSolver.affectPosition) {
                position.copyFrom(this.appendTransformSolver.appendPosition);
            }
        }

        if (this.ikChainInfo !== null) {
            this.ikChainInfo.localRotation.copyFrom(rotation);
            this.ikChainInfo.localPosition.copyFrom(position);

            this.ikChainInfo.ikRotation.multiplyToRef(rotation, rotation);
        }

        const worldMatrix = MmdRuntimeBone._TempMatrix;

        const localPosition = this.linkedBone.getRestMatrix().getTranslationToRef(MmdRuntimeBone._TempPosition2)
            .addInPlace(position);

        const scaling = this.linkedBone.scaling;
        if (scaling.x !== 1 || scaling.y !== 1 || scaling.z !== 1) {
            Matrix.ComposeToRef(scaling, rotation, localPosition, worldMatrix);
        } else {
            Matrix.FromQuaternionToRef(rotation, worldMatrix);
            worldMatrix.setTranslation(localPosition);
        }

        if (this.parentBone !== null) {
            const parentWorldMatrix = this.parentBone.getWorldMatrixToRef(MmdRuntimeBone._TempMatrix2);
            worldMatrix.multiplyToRef(parentWorldMatrix, worldMatrix);
        }

        worldMatrix.copyToArray(this.worldMatrix);

        if (computeIk && this.ikSolver !== null) {
            if (!(usePhysics && this.ikSolver.canSkipWhenPhysicsEnabled)) {
                this.ikSolver.solve(usePhysics);
            }
        }
    }

    /**
     * @internal
     */
    public updateIkChainWorldMatrix(): void {
        const ikChainInfo = this.ikChainInfo!;

        const rotation = ikChainInfo.ikRotation.multiplyToRef(ikChainInfo.localRotation, MmdRuntimeBone._TempRotation);

        const worldMatrix = MmdRuntimeBone._TempMatrix;

        const scaling = this.linkedBone.scaling;
        if (scaling.x !== 1 || scaling.y !== 1 || scaling.z !== 1) {
            Matrix.ScalingToRef(scaling.x, scaling.y, scaling.z, worldMatrix);
            const rotationMatrix = Matrix.FromQuaternionToRef(rotation, MmdRuntimeBone._TempMatrix2);
            worldMatrix.multiplyToRef(rotationMatrix, worldMatrix);
        } else {
            Matrix.FromQuaternionToRef(rotation, worldMatrix);
        }

        const localPosition = this.linkedBone.getRestMatrix().getTranslationToRef(MmdRuntimeBone._TempPosition);
        localPosition.addInPlace(ikChainInfo.localPosition);
        worldMatrix.setTranslation(localPosition);

        if (this.parentBone !== null) {
            const parentWorldMatrix = this.parentBone.getWorldMatrixToRef(MmdRuntimeBone._TempMatrix2);
            worldMatrix.multiplyToRef(parentWorldMatrix, worldMatrix);
        }

        worldMatrix.copyToArray(this.worldMatrix);

        const childBones = this.childBones;
        for (let i = 0; i < childBones.length; ++i) {
            childBones[i]._updateWorldMatrixRecursive();
        }
    }

    private static readonly _WorldMatrixUpdateStack: MmdRuntimeBone[] = [];

    private _updateWorldMatrixRecursive(): void {
        const stack = MmdRuntimeBone._WorldMatrixUpdateStack;
        stack.length = 0;

        stack.push(this);
        while (stack.length > 0) {
            const bone = stack.pop()!;
            bone.updateWorldMatrix(false, false);

            for (let i = 0, n = bone.childBones.length; i < n; ++i) {
                stack.push(bone.childBones[i]);
            }
        }
    }

    /**
     * Get the world matrix of this bone
     *
     * The result of this method is not same as `linkedBone.getFinalMatrix()`
     *
     * `linkedBone.getFinalMatrix()` updated at the end of the mmd runtime update, so it may not be the latest value
     * @param target target matrix
     * @returns target matrix
     */
    public getWorldMatrixToRef(target: Matrix): Matrix {
        return Matrix.FromArrayToRef(this.worldMatrix, 0, target);
    }

    /**
     * Get the world translation of this bone
     * @param target target vector
     * @returns target vector
     */
    public getWorldTranslationToRef(target: Vector3): Vector3 {
        return Vector3.FromArrayToRef(this.worldMatrix, 12, target);
    }

    /**
     * Set the world translation of this bone
     * @param source source vector
     */
    public setWorldTranslationFromRef(source: Vector3): void {
        source.toArray(this.worldMatrix, 12);
    }

    /**
     * Get ik solver index
     *
     * If the bone does not have an ik solver, it will return -1
     */
    public get ikSolverIndex(): number {
        return this.ikSolver !== null ? this.ikSolver.index : -1;
    }
}

File: src/Runtime/Optimized/wasm_src/src/mmd_model/mmd_runtime_bone.rs
use glam::{Vec3, Vec3A, Mat4, Quat};

use crate::unchecked_slice::{UncheckedSlice, UncheckedSliceMut};

use super::{animation_arena::AnimationArena, ik_chain_info::IkChainInfo, MmdModel};

pub(crate) struct MmdRuntimeBoneArena {
    arena: Box<[MmdRuntimeBone]>,
    world_matrix_arena: Box<[Mat4]>,
    world_matrix_back_buffer: Option<Box<[Mat4]>>,
}

impl MmdRuntimeBoneArena {
    pub(super) fn new(arena: Box<[MmdRuntimeBone]>) -> Self {
        let bone_count = arena.len();
        MmdRuntimeBoneArena {
            arena,
            world_matrix_arena: vec![Mat4::IDENTITY; bone_count].into_boxed_slice(),
            world_matrix_back_buffer: None,
        }
    }
    
    #[inline]
    pub(crate) fn arena(&self) -> UncheckedSlice<MmdRuntimeBone> {
        UncheckedSlice::new(&self.arena)
    }

    #[inline]
    pub(crate) fn arena_mut(&mut self) -> UncheckedSliceMut<MmdRuntimeBone> {
        UncheckedSliceMut::new(&mut self.arena)
    }

    #[inline]
    pub(crate) fn world_matrix_arena_mut_ptr(&mut self) -> *mut f32 {
        self.world_matrix_arena.as_mut_ptr() as *mut f32
    }

    pub(crate) fn create_world_matrix_back_buffer(&mut self) -> *mut f32 {
        self.world_matrix_back_buffer = Some(vec![Mat4::IDENTITY; self.world_matrix_arena.len()].into_boxed_slice());
        self.world_matrix_back_buffer.as_mut().unwrap().as_mut_ptr() as *mut f32
    }

    pub(crate) fn swap_buffer(&mut self) {
        let mut back_buffer = self.world_matrix_back_buffer.take().unwrap();
        std::mem::swap(&mut self.world_matrix_arena, &mut back_buffer);
        self.world_matrix_back_buffer = Some(back_buffer);
    }

    #[inline]
    pub(crate) fn world_matrices(&self) -> UncheckedSlice<Mat4> {
        UncheckedSlice::new(&self.world_matrix_arena)
    }

    #[inline]
    pub(crate) fn world_matrices_mut(&mut self) -> UncheckedSliceMut<Mat4> {
        UncheckedSliceMut::new(&mut self.world_matrix_arena)
    }

    pub(super) fn reset_world_matrices(&mut self) {
        for world_matrix in self.world_matrices_mut().iter_mut() {
            *world_matrix = Mat4::IDENTITY;
        }
    }
}

pub(crate) struct MmdRuntimeBone {
    pub(super) rest_position: Vec3A,
    pub(super) absolute_inverse_bind_matrix: Mat4,
    index: u32,

    pub(super) parent_bone: Option<u32>,
    pub(super) child_bones: Vec<u32>,
    pub(super) transform_order: i32,
    pub(super) transform_after_physics: bool,

    pub(super) append_transform_solver: Option<u32>,
    pub(super) axis_limit: Option<Vec3>,
    pub(super) ik_solver: Option<u32>,

    pub(super) morph_position_offset: Option<Vec3A>,
    pub(super) morph_rotation_offset: Option<Quat>,

    pub(super) ik_chain_info: Option<IkChainInfo>,

    pub(super) rigidbody_indices: Box<[u32]>,
}

impl MmdRuntimeBone {
    pub(super) fn new(index: u32) -> Self {
        MmdRuntimeBone {
            rest_position: Vec3A::ZERO,
            absolute_inverse_bind_matrix: Mat4::IDENTITY,
            index,
            
            parent_bone: None,
            child_bones: Vec::new(),
            transform_order: 0,
            transform_after_physics: false,

            append_transform_solver: None,
            axis_limit: None,
            ik_solver: None,

            morph_position_offset: None,
            morph_rotation_offset: None,

            ik_chain_info: None,

            rigidbody_indices: Box::new([]),
        }
    }

    #[inline]
    pub(crate) fn rest_position(&self) -> Vec3A {
        self.rest_position
    }

    #[inline]
    pub(crate) fn absolute_inverse_bind_matrix(&self) -> &Mat4 {
        &self.absolute_inverse_bind_matrix
    }

    #[inline]
    pub(crate) fn parent_bone(&self) -> Option<u32> {
        self.parent_bone
    }

    #[inline]
    #[allow(dead_code)]
    pub(crate) fn rigidbody_indices(&self) -> &[u32] {
        &self.rigidbody_indices
    }

    pub(super) fn animated_position(&self, animation_arena: &AnimationArena) -> Vec3A {
        let mut position = animation_arena.bone_arena()[self.index].position;
        if let Some(morph_position_offset) = self.morph_position_offset {
            position += morph_position_offset;
        }
        position
    }

    pub(super) fn animated_rotation(&self, animation_arena: &AnimationArena) -> Quat {
        let mut rotation = animation_arena.bone_arena()[self.index].rotation;
        
        // MMD's implementation transforms the rotation axis to fit the axis limit of the target skeleton at animation load time.
        // However, that method makes it impossible to apply one animation data to multiple models,
        // so we use an implementation that performs the axis transformation at runtime.
        if let Some(axis_limit) = self.axis_limit {
            if axis_limit == Vec3::ZERO {
                return Quat::IDENTITY;
            }

            let (animation_axis, mut angle) = rotation.to_axis_angle();
            if animation_axis.dot(axis_limit) < 0.0 {
                angle = -angle;
            }
            rotation = Quat::from_axis_angle(axis_limit, angle);
        }

        if let Some(morph_rotation_offset) = self.morph_rotation_offset {
            rotation = morph_rotation_offset * rotation;
        }
        rotation
    }

    #[inline]
    pub(super) fn animation_position_offset(&self, animation_arena: &AnimationArena) -> Vec3A {
        self.animated_position(animation_arena) - self.rest_position
    }

    pub(super) fn reset_state(&mut self) {
        if let Some(ik_chain_info) = &mut self.ik_chain_info {
            ik_chain_info.reset_state();
        }
    }
}

impl MmdModel {
    pub(super) fn update_world_matrix(&mut self, bone_index: u32, use_physics: bool, compute_ik: bool) {
        let bone = &self.bone_arena.arena()[bone_index];

        let mut rotation = bone.animated_rotation(&self.animation_arena);
        let mut position = bone.animation_position_offset(&self.animation_arena);

        if let Some(append_transform_solver) = bone.append_transform_solver {
            self.append_transform_solver_arena.update(
                append_transform_solver,
                &self.animation_arena,
                &mut self.bone_arena,
                rotation,
                position
            );
            let append_transform_solver = &self.append_transform_solver_arena.arena()[append_transform_solver];

            if append_transform_solver.is_affect_rotation() {
                rotation = append_transform_solver.append_rotation();
            }
            if append_transform_solver.is_affect_position() {
                position = append_transform_solver.append_position();
            }
        }

        let bone = &mut self.bone_arena.arena_mut()[bone_index];
        
        if let Some(ik_chain_info) = &mut bone.ik_chain_info {
            *ik_chain_info.local_rotation_mut() = rotation;
            *ik_chain_info.local_position_mut() = position;
            
            rotation = ik_chain_info.ik_rotation() * rotation;
        }

        let local_scale = self.animation_arena.bone_arena()[bone_index].scale;
        let local_position = position + bone.rest_position;

        let local_matrix = if local_scale.x != 1.0 || local_scale.y != 1.0 || local_scale.z != 1.0 {
            Mat4::from_scale_rotation_translation(local_scale.into(), rotation, local_position.into())
        } else {
            Mat4::from_rotation_translation(rotation, local_position.into())
        };

        let world_matrix = if let Some(parent_bone) = bone.parent_bone {
            let parent_world_matrix = self.bone_arena.world_matrices()[parent_bone];
            parent_world_matrix * local_matrix
        } else {
            local_matrix
        };
        
        self.bone_arena.world_matrices_mut()[bone_index] = world_matrix;

        let bone = &mut self.bone_arena.arena_mut()[bone_index];

        if compute_ik {
            if let Some(ik_solver_index) = bone.ik_solver {
                let ik_solver = &self.ik_solver_arena.arena()[ik_solver_index];
                if !(use_physics && ik_solver.can_skip_when_physics_enabled()) {
                    self.solve_ik(ik_solver_index, use_physics);
                }
            }
        }
    }

    pub(super) fn update_ik_chain_world_matrix(&mut self, bone_index: u32) {
        let bone = &self.bone_arena.arena()[bone_index];
        let ik_chain_info = bone.ik_chain_info.as_ref().unwrap();

        let rotation = ik_chain_info.ik_rotation() * ik_chain_info.local_rotation();

        let local_scale = self.animation_arena.bone_arena()[bone_index].scale;
        let local_position = ik_chain_info.local_position() + bone.rest_position;

        let local_matrix = if local_scale.x != 1.0 || local_scale.y != 1.0 || local_scale.z != 1.0 {
            Mat4::from_scale_rotation_translation(local_scale.into(), rotation, local_position.into())
        } else {
            Mat4::from_rotation_translation(rotation, local_position.into())
        };

        let world_matrix = if let Some(parent_bone) = bone.parent_bone {
            let parent_world_matrix = self.bone_arena.world_matrices()[parent_bone];
            parent_world_matrix * local_matrix
        } else {
            local_matrix
        };
        
        self.bone_arena.world_matrices_mut()[bone_index] = world_matrix;

        let bone = &self.bone_arena.arena()[bone_index];
        for i in 0..bone.child_bones.len() {
            let child_bone = self.bone_arena.arena()[bone_index].child_bones[i];
            self.update_world_matrix_recursive(child_bone);
        }
    }

    fn update_world_matrix_recursive(&mut self, bone_index: u32) {
        let mut bone_stack = self.bone_stack.take().unwrap();
        bone_stack.clear();

        bone_stack.push(bone_index);
        while let Some(bone_index) = bone_stack.pop() {
            self.update_world_matrix(bone_index, false, false);

            let bone = &self.bone_arena().arena()[bone_index];
            for &child_bone in bone.child_bones.iter() {
                bone_stack.push(child_bone);
            }
        }

        self.bone_stack = Some(bone_stack);
    }
}
